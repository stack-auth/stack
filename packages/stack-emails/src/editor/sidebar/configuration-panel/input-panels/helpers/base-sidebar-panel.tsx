import { SimpleTooltip, Typography } from '@stackframe/stack-ui';
import React from 'react';

type SidebarPanelProps = {
  title: string,
  children: React.ReactNode,
  tooltip?: string,
};
export default function BaseSidebarPanel({ title, children, tooltip }: SidebarPanelProps) {
  return (
    <div className='py-2 px-4'>
      <div className='flex items-center mb-6 gap-2'>
        <Typography variant='secondary' className='font-medium'>
          {title}
        </Typography>
        {tooltip && <SimpleTooltip tooltip={tooltip} type='info'/>}
      </div>
      <div className='flex flex-col gap-6 mb-3'>
        {children}
      </div>
    </div>
  );
}
