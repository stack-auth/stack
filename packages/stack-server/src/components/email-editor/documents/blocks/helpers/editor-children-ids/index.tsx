import React, { Fragment } from 'react';

import { TEditorBlock } from '../../../editor/core';
import EditorBlock from '../../../editor/editor-block';

import AddBlockButton from './add-block-menu';
import { useSelectedBlockId } from '../../../editor/editor-context';

export type EditorChildrenChange = {
  blockId: string,
  block: TEditorBlock,
  childrenIds: string[],
};

function generateId() {
  return `block-${Date.now()}`;
}

export type EditorChildrenIdsProps = {
  childrenIds: string[] | null | undefined,
  onChange: (val: EditorChildrenChange) => void,
};
export default function EditorChildrenIds({ childrenIds, onChange }: EditorChildrenIdsProps) {
  const selectedBlockId = useSelectedBlockId();

  const appendBlock = (block: TEditorBlock) => {
    const blockId = generateId();
    return onChange({
      blockId,
      block,
      childrenIds: [...(childrenIds || []), blockId],
    });
  };

  const insertBlock = (block: TEditorBlock, index: number) => {
    const blockId = generateId();
    const newChildrenIds = [...(childrenIds || [])];
    newChildrenIds.splice(index, 0, blockId);
    return onChange({
      blockId,
      block,
      childrenIds: newChildrenIds,
    });
  };

  if (!childrenIds || childrenIds.length === 0) {
    return <AddBlockButton placeholder onSelect={appendBlock} />;
  }

  return (
    <>
      {childrenIds.map((childId, i) => (
        <Fragment key={childId}>
          {childId === selectedBlockId && <AddBlockButton onSelect={(block) => insertBlock(block, i)} />}
          <EditorBlock id={childId} />
          {childId === selectedBlockId && <AddBlockButton onSelect={(block) => insertBlock(block, i + 1)} />}
        </Fragment>
      ))}
    </>
  );
}
