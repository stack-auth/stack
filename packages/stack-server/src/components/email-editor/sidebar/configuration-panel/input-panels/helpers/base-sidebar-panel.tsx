import React from 'react';

import { Typography } from '@mui/material';
import { SimpleTooltip } from '@/components/simple-tooltip';

type SidebarPanelProps = {
  title: string,
  children: React.ReactNode,
  tooltip?: string,
};
export default function BaseSidebarPanel({ title, children, tooltip }: SidebarPanelProps) {
  return (
    <div className='py-2 px-4'>
      <div className='flex items-center mb-2 gap-2'>
        <Typography variant="overline" color="text.secondary">
          {title}
        </Typography>
        {tooltip && <SimpleTooltip tooltip={tooltip} type='info'/>}
      </div>
      <div className='flex flex-col space-y-5 mb-3'>
        {children}
      </div>
    </div>
  );
}
