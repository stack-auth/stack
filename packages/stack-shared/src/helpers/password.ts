import { KnownErrors } from "..";

const minLength = 8;
const maxLength = 256;

export function getPasswordError(password: string): KnownErrors["PasswordRequirementsNotMet"] | undefined {
  if (password.length < minLength) {
    return new KnownErrors.PasswordTooShort(minLength);
  }

  if (password.length > maxLength) {
    return new KnownErrors.PasswordTooLong(maxLength);
  }

  return undefined;
}
