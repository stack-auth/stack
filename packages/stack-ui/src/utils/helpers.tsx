import { ThemeConfig } from '../providers/ui-provider';

function mergeDefaultTheme(theme: ThemeConfig, defaultTheme: ThemeConfig) {
}