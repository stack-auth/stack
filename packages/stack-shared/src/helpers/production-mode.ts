import { ProjectsCrud } from "../interface/crud/projects";
import { StackAssertionError } from "../utils/errors";
import { isLocalhost } from "../utils/urls";

export type ProductionModeError = {
  message: string,
  relativeFixUrl: `/${string}`,
};

export function getProductionModeErrors(project: ProjectsCrud["Admin"]["Read"]): ProductionModeError[] {
  const errors: ProductionModeError[] = [];
  const domainsFixUrl = `/projects/${project.id}/domains` as const;

  if (project.config.allow_localhost) {
    errors.push({
      message: "Localhost is not allowed in production mode, turn off 'Allow localhost' in project settings",
      relativeFixUrl: domainsFixUrl,
    });
  }

  for (const { domain } of project.config.domains) {
    let url;
    try {
      url = new URL(domain);
    } catch (e) {
      throw new StackAssertionError("Domain was somehow not a valid URL; we should've caught this when setting the domain in the first place", {
        domain,
        projectId: project
      });
    }

    if (isLocalhost(url)) {
      errors.push({
        message: "Localhost domains are not allowed to be trusted in production mode: " + domain,
        relativeFixUrl: domainsFixUrl,
      });
    } else if (url.hostname.match(/^\d+(\.\d+)*$/)) {
      errors.push({
        message: "Direct IPs are not valid for trusted domains in production mode: " + domain,
        relativeFixUrl: domainsFixUrl,
      });
    } else if (url.protocol !== "https:") {
      errors.push({
        message: "Trusted domains should be HTTPS: " + domain,
        relativeFixUrl: domainsFixUrl,
      });
    }
  }

  return errors;
}