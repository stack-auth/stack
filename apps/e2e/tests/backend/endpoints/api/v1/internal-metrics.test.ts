import { it } from "../../../../helpers";
import { Auth, Project, backendContext, createMailbox, niceBackendFetch } from "../../../backend-helpers";

it("should return metrics data", async ({ expect }) => {
  await Project.createAndSwitch({
    config: {
      magic_link_enabled: true,
    }
  });

  const response = await niceBackendFetch("/api/v1/internal/metrics", { accessType: 'admin' });
  expect(response).toMatchSnapshot(`metrics_result_no_users`);
});

it("should return metrics data with users", async ({ expect }) => {
  await Project.createAndSwitch({
    config: {
      magic_link_enabled: true,
    }
  });

  const mailboxes = new Array(10).fill(null).map(() => createMailbox());

  backendContext.set({ mailbox: mailboxes[0], ipData: { country: "AQ", ipAddress: "127.0.0.1", city: "[placeholder city]", region: "NQ", latitude: 68, longitude: 30, tzIdentifier: "Europe/Zurich" } });
  await Auth.Otp.signIn();

  for (const mailbox of mailboxes) {
    backendContext.set({ mailbox, ipData: undefined });
    await Auth.Otp.signIn();
  }
  backendContext.set({ mailbox: mailboxes[8] });
  await Auth.Otp.signIn();
  const deleteResponse = await niceBackendFetch("/api/v1/users/me", {
    accessType: "server",
    method: "DELETE",
  });
  expect(deleteResponse.status).toBe(200);
  backendContext.set({ userAuth: { ...backendContext.value.userAuth, accessToken: undefined } });

  backendContext.set({ mailbox: mailboxes[1], ipData: { country: "CH", ipAddress: "127.0.0.1", city: "Zurich", region: "ZH", latitude: 47.3769, longitude: 8.5417, tzIdentifier: "Europe/Zurich" } });
  await Auth.Otp.signIn();
  backendContext.set({ mailbox: mailboxes[1], ipData: { country: "AQ", ipAddress: "127.0.0.1", city: "[placeholder city]", region: "NQ", latitude: 68, longitude: 30, tzIdentifier: "Europe/Zurich" } });
  await Auth.Otp.signIn();
  backendContext.set({ mailbox: mailboxes[2], ipData: { country: "CH", ipAddress: "127.0.0.1", city: "Zurich", region: "ZH", latitude: 47.3769, longitude: 8.5417, tzIdentifier: "Europe/Zurich" } });
  await Auth.Otp.signIn();

  const response = await niceBackendFetch("/api/v1/internal/metrics", { accessType: 'admin' });
  expect(response).toMatchSnapshot();
});

it("should not work for non-admins", async ({ expect }) => {
  await Project.createAndSwitch({
    config: {
      magic_link_enabled: true,
    }
  });

  await Auth.Otp.signIn();

  const response = await niceBackendFetch("/api/v1/internal/metrics", { accessType: 'server' });
  expect(response).toMatchInlineSnapshot(`
    NiceResponse {
      "status": 401,
      "body": {
        "code": "INSUFFICIENT_ACCESS_TYPE",
        "details": {
          "actual_access_type": "server",
          "allowed_access_types": ["admin"],
        },
        "error": "The x-stack-access-type header must be 'admin', but was 'server'.",
      },
      "headers": Headers {
        "x-stack-known-error": "INSUFFICIENT_ACCESS_TYPE",
        <some fields may have been hidden>,
      },
    }
  `);
});
