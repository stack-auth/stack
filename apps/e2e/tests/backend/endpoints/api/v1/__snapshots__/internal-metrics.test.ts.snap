// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should return metrics data > metrics_result_no_users 1`] = `
NiceResponse {
  "status": 200,
  "body": {
    "daily_active_users": [
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
    ],
    "daily_users": [
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
    ],
    "login_methods": [],
    "recently_active": [],
    "recently_registered": [],
    "total_users": 0,
    "users_by_country": {},
  },
  "headers": Headers { <some fields may have been hidden> },
}
`;

exports[`should return metrics data with users 17`] = `
NiceResponse {
  "status": 200,
  "body": {
    "daily_active_users": [
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 10,
        "date": <stripped field 'date'>,
      },
    ],
    "daily_users": [
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 0,
        "date": <stripped field 'date'>,
      },
      {
        "activity": 9,
        "date": <stripped field 'date'>,
      },
    ],
    "login_methods": [
      {
        "count": 9,
        "method": "otp",
      },
    ],
    "recently_active": [
      {
        "auth_with_email": true,
        "client_metadata": null,
        "client_read_only_metadata": null,
        "display_name": null,
        "has_password": false,
        "id": "<stripped UUID>",
        "last_active_at_millis": <stripped field 'last_active_at_millis'>,
        "oauth_providers": [],
        "otp_auth_enabled": true,
        "passkey_auth_enabled": false,
        "primary_email": "mailbox-3--<stripped UUID>@stack-generated.example.com",
        "primary_email_auth_enabled": true,
        "primary_email_verified": true,
        "profile_image_url": null,
        "requires_totp_mfa": false,
        "selected_team": null,
        "selected_team_id": null,
        "server_metadata": null,
        "signed_up_at_millis": <stripped field 'signed_up_at_millis'>,
      },
      {
        "auth_with_email": true,
        "client_metadata": null,
        "client_read_only_metadata": null,
        "display_name": null,
        "has_password": false,
        "id": "<stripped UUID>",
        "last_active_at_millis": <stripped field 'last_active_at_millis'>,
        "oauth_providers": [],
        "otp_auth_enabled": true,
        "passkey_auth_enabled": false,
        "primary_email": "mailbox-2--<stripped UUID>@stack-generated.example.com",
        "primary_email_auth_enabled": true,
        "primary_email_verified": true,
        "profile_image_url": null,
        "requires_totp_mfa": false,
        "selected_team": null,
        "selected_team_id": null,
        "server_metadata": null,
        "signed_up_at_millis": <stripped field 'signed_up_at_millis'>,
      },
      {
        "auth_with_email": true,
        "client_metadata": null,
        "client_read_only_metadata": null,
        "display_name": null,
        "has_password": false,
        "id": "<stripped UUID>",
        "last_active_at_millis": <stripped field 'last_active_at_millis'>,
        "oauth_providers": [],
        "otp_auth_enabled": true,
        "passkey_auth_enabled": false,
        "primary_email": "mailbox-10--<stripped UUID>@stack-generated.example.com",
        "primary_email_auth_enabled": true,
        "primary_email_verified": true,
        "profile_image_url": null,
        "requires_totp_mfa": false,
        "selected_team": null,
        "selected_team_id": null,
        "server_metadata": null,
        "signed_up_at_millis": <stripped field 'signed_up_at_millis'>,
      },
      {
        "auth_with_email": true,
        "client_metadata": null,
        "client_read_only_metadata": null,
        "display_name": null,
        "has_password": false,
        "id": "<stripped UUID>",
        "last_active_at_millis": <stripped field 'last_active_at_millis'>,
        "oauth_providers": [],
        "otp_auth_enabled": true,
        "passkey_auth_enabled": false,
        "primary_email": "mailbox-8--<stripped UUID>@stack-generated.example.com",
        "primary_email_auth_enabled": true,
        "primary_email_verified": true,
        "profile_image_url": null,
        "requires_totp_mfa": false,
        "selected_team": null,
        "selected_team_id": null,
        "server_metadata": null,
        "signed_up_at_millis": <stripped field 'signed_up_at_millis'>,
      },
    ],
    "recently_registered": [
      {
        "auth_with_email": true,
        "client_metadata": null,
        "client_read_only_metadata": null,
        "display_name": null,
        "has_password": false,
        "id": "<stripped UUID>",
        "last_active_at_millis": <stripped field 'last_active_at_millis'>,
        "oauth_providers": [],
        "otp_auth_enabled": true,
        "passkey_auth_enabled": false,
        "primary_email": "mailbox-10--<stripped UUID>@stack-generated.example.com",
        "primary_email_auth_enabled": true,
        "primary_email_verified": true,
        "profile_image_url": null,
        "requires_totp_mfa": false,
        "selected_team": null,
        "selected_team_id": null,
        "server_metadata": null,
        "signed_up_at_millis": <stripped field 'signed_up_at_millis'>,
      },
      {
        "auth_with_email": true,
        "client_metadata": null,
        "client_read_only_metadata": null,
        "display_name": null,
        "has_password": false,
        "id": "<stripped UUID>",
        "last_active_at_millis": <stripped field 'last_active_at_millis'>,
        "oauth_providers": [],
        "otp_auth_enabled": true,
        "passkey_auth_enabled": false,
        "primary_email": "mailbox-8--<stripped UUID>@stack-generated.example.com",
        "primary_email_auth_enabled": true,
        "primary_email_verified": true,
        "profile_image_url": null,
        "requires_totp_mfa": false,
        "selected_team": null,
        "selected_team_id": null,
        "server_metadata": null,
        "signed_up_at_millis": <stripped field 'signed_up_at_millis'>,
      },
      {
        "auth_with_email": true,
        "client_metadata": null,
        "client_read_only_metadata": null,
        "display_name": null,
        "has_password": false,
        "id": "<stripped UUID>",
        "last_active_at_millis": <stripped field 'last_active_at_millis'>,
        "oauth_providers": [],
        "otp_auth_enabled": true,
        "passkey_auth_enabled": false,
        "primary_email": "mailbox-7--<stripped UUID>@stack-generated.example.com",
        "primary_email_auth_enabled": true,
        "primary_email_verified": true,
        "profile_image_url": null,
        "requires_totp_mfa": false,
        "selected_team": null,
        "selected_team_id": null,
        "server_metadata": null,
        "signed_up_at_millis": <stripped field 'signed_up_at_millis'>,
      },
      {
        "auth_with_email": true,
        "client_metadata": null,
        "client_read_only_metadata": null,
        "display_name": null,
        "has_password": false,
        "id": "<stripped UUID>",
        "last_active_at_millis": <stripped field 'last_active_at_millis'>,
        "oauth_providers": [],
        "otp_auth_enabled": true,
        "passkey_auth_enabled": false,
        "primary_email": "mailbox-6--<stripped UUID>@stack-generated.example.com",
        "primary_email_auth_enabled": true,
        "primary_email_verified": true,
        "profile_image_url": null,
        "requires_totp_mfa": false,
        "selected_team": null,
        "selected_team_id": null,
        "server_metadata": null,
        "signed_up_at_millis": <stripped field 'signed_up_at_millis'>,
      },
      {
        "auth_with_email": true,
        "client_metadata": null,
        "client_read_only_metadata": null,
        "display_name": null,
        "has_password": false,
        "id": "<stripped UUID>",
        "last_active_at_millis": <stripped field 'last_active_at_millis'>,
        "oauth_providers": [],
        "otp_auth_enabled": true,
        "passkey_auth_enabled": false,
        "primary_email": "mailbox-5--<stripped UUID>@stack-generated.example.com",
        "primary_email_auth_enabled": true,
        "primary_email_verified": true,
        "profile_image_url": null,
        "requires_totp_mfa": false,
        "selected_team": null,
        "selected_team_id": null,
        "server_metadata": null,
        "signed_up_at_millis": <stripped field 'signed_up_at_millis'>,
      },
    ],
    "total_users": 9,
    "users_by_country": {
      "AQ": 2,
      "CH": 1,
    },
  },
  "headers": Headers { <some fields may have been hidden> },
}
`;
